[Lens] Formula: add validation for multiple field/metrics

dej611 opened the pull request:
## Summary

This PR adds one more validation case if the user passes too many fields/metrics to an operation.

<img width="341" alt="Screenshot 2021-07-01 at 14 38 58" src="https://user-images.githubusercontent.com/924948/124126380-ea7f2b00-da7a-11eb-99bf-944ca2fca4b1.png">
<img width="355" alt="Screenshot 2021-07-01 at 14 39 17" src="https://user-images.githubusercontent.com/924948/124126386-ebb05800-da7a-11eb-97b3-17351199b189.png">
<img width="354" alt="Screenshot 2021-07-01 at 14 39 32" src="https://user-images.githubusercontent.com/924948/124126389-ec48ee80-da7a-11eb-9dc6-ee6e9a946191.png">


### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/master/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

elasticmachine commented:
Pinging @elastic/kibana-app (Team:KibanaApp)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/)
* Commit: 0dcc0b83ead8cc942efa11b41bfe4b1c549794bd
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104092/0dcc0b83ead8cc942efa11b41bfe4b1c549794bd)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/6e3df60abaf1bcf1cc0ea902dde3913820fc32c8) | [after](https://github.com/elastic/kibana/commit/0dcc0b83ead8cc942efa11b41bfe4b1c549794bd) | diff |
| --- | --- | --- | --- |
| `lens` | 1.5MB | 1.5MB | +912.0B |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/","number":"135961","commit":"0dcc0b83ead8cc942efa11b41bfe4b1c549794bd"}]}
PIPELINE-->

wylieconlon reviewed (changes requested):
While this logic appears to work for cases where the user has duplicate types, like `sum(bytes, bytes)` or `moving_average(count(), sum())`, it does not catch mixed types like `sum(bytes, count())` which are also invalid.

Also, I think we should backport this 7.14 and mark this as a `release_note:skip`, I would mark it as a bugfix only if we ship the broken state to users.

dej611 commented:
Add one more validation case, for the mixed mode in particular, and enriched tests with different type of parameters as well.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/)
* Commit: b50ecd55e0124313ad62b1c18db48cb62ed01852
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104092/b50ecd55e0124313ad62b1c18db48cb62ed01852)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/6e3df60abaf1bcf1cc0ea902dde3913820fc32c8) | [after](https://github.com/elastic/kibana/commit/b50ecd55e0124313ad62b1c18db48cb62ed01852) | diff |
| --- | --- | --- | --- |
| `lens` | 1.5MB | 1.5MB | +1.7KB |

### History
* :green_heart: [Build #135961](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/) succeeded 0dcc0b83ead8cc942efa11b41bfe4b1c549794bd

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/","number":"135961","commit":"0dcc0b83ead8cc942efa11b41bfe4b1c549794bd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/","number":"136221","commit":"b50ecd55e0124313ad62b1c18db48cb62ed01852"}]}
PIPELINE-->

flash1293 commented:
@elasticmachine merge upstream

flash1293 commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/formula/validation.ts:
```diff
@@ -633,6 +701,22 @@ function runFullASTValidation(
             })
           );
         } else {
+          const fields = variables.filter(
```
Which case is caught here?

dej611 replied:
Probably need to add some more comment context on all these branches.
In this specific branch it is checking that only a single `field` is passed to `field`-only operations.

flash1293 replied:
I don't think so, this branch is in the `if (nodeOperation.input === 'fullReference')` block.

I don't see in which case it would become necessary - the only one is this, but it shows two error for the same problem:
<img width="708" alt="Screenshot 2021-07-06 at 11 55 05" src="https://user-images.githubusercontent.com/1508364/124581179-2803ff80-de51-11eb-9752-456b2fb6ab3a.png">

dej611 replied:
You're right sorry. The check has to exclude the first argument.
This check is here to validate this scenario:

```
moving_average(average(bytes), bytes)
```

I'll refactor the logic to group both first argument check and all arguents check for field values.

flash1293 commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/formula/validation.ts:
```diff
@@ -621,6 +674,21 @@ function runFullASTValidation(
             })
           );
+        } else {
+          if (functions.length > 1) {
```
I'm not sure whether this is a future-proof way to check - the way our full reference operations are set up, they could theoretically accept multiple parameters. Should we check the `requiredReferences` here?

dej611 replied:
This code is generic enough to check functions in different position in the node, as `moving_average( median( ...), median( ... ) )` as well as `moving_average( median( ...), window = 7, shift="7d", median( ... ) )`.

I could parametrize the threshold to be `requiredReferences.filter( /* check for functions */).length` here, but in general the formula code assumes only one argument to be a reference (to a field or operation) and it may be required to change in the future if we introduce operations with different assumptions.
The parametric change here is pretty easy and I'll push a fix on that.

flash1293 reviewed (commented):
Didn't manage to break it, but I'm unsure about some things - left some questions.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/)
* Commit: 13eb57cfdeb019aaa92ef8162abe30cd64b960ba
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104092/13eb57cfdeb019aaa92ef8162abe30cd64b960ba)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/7395fee0f9b3d363979a3bcf7f668898b26c3fa8) | [after](https://github.com/elastic/kibana/commit/13eb57cfdeb019aaa92ef8162abe30cd64b960ba) | diff |
| --- | --- | --- | --- |
| `lens` | 1.5MB | 1.5MB | +1.7KB |

### History
* :green_heart: [Build #136221](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/) succeeded b50ecd55e0124313ad62b1c18db48cb62ed01852
* :green_heart: [Build #135961](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/) succeeded 0dcc0b83ead8cc942efa11b41bfe4b1c549794bd

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/","number":"135961","commit":"0dcc0b83ead8cc942efa11b41bfe4b1c549794bd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/","number":"136221","commit":"b50ecd55e0124313ad62b1c18db48cb62ed01852"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/","number":"136259","commit":"13eb57cfdeb019aaa92ef8162abe30cd64b960ba"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/)
* Commit: 950ff349365524c1982c14fe062a492af499d729
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104092/950ff349365524c1982c14fe062a492af499d729)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/execution/node/190/log)

### History
* :green_heart: [Build #136259](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/) succeeded 13eb57cfdeb019aaa92ef8162abe30cd64b960ba
* :green_heart: [Build #136221](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/) succeeded b50ecd55e0124313ad62b1c18db48cb62ed01852
* :green_heart: [Build #135961](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/) succeeded 0dcc0b83ead8cc942efa11b41bfe4b1c549794bd

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/","number":"135961","commit":"0dcc0b83ead8cc942efa11b41bfe4b1c549794bd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/","number":"136221","commit":"b50ecd55e0124313ad62b1c18db48cb62ed01852"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/","number":"136259","commit":"13eb57cfdeb019aaa92ef8162abe30cd64b960ba"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/","number":"136294","commit":"950ff349365524c1982c14fe062a492af499d729"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/)
* Commit: 974808337015f94ec1cc9c8e2422b8b77c5ac7bd
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104092/974808337015f94ec1cc9c8e2422b8b77c5ac7bd)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/7395fee0f9b3d363979a3bcf7f668898b26c3fa8) | [after](https://github.com/elastic/kibana/commit/974808337015f94ec1cc9c8e2422b8b77c5ac7bd) | diff |
| --- | --- | --- | --- |
| `lens` | 1.5MB | 1.5MB | +1.8KB |

### History
* :broken_heart: [Build #136294](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/) failed 950ff349365524c1982c14fe062a492af499d729
* :green_heart: [Build #136259](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/) succeeded 13eb57cfdeb019aaa92ef8162abe30cd64b960ba
* :green_heart: [Build #136221](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/) succeeded b50ecd55e0124313ad62b1c18db48cb62ed01852
* :green_heart: [Build #135961](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/) succeeded 0dcc0b83ead8cc942efa11b41bfe4b1c549794bd

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/","number":"135961","commit":"0dcc0b83ead8cc942efa11b41bfe4b1c549794bd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/","number":"136221","commit":"b50ecd55e0124313ad62b1c18db48cb62ed01852"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/","number":"136259","commit":"13eb57cfdeb019aaa92ef8162abe30cd64b960ba"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/","number":"136294","commit":"950ff349365524c1982c14fe062a492af499d729"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/","number":"136321","commit":"974808337015f94ec1cc9c8e2422b8b77c5ac7bd"}]}
PIPELINE-->

flash1293 commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/formula/validation.ts:
```diff
@@ -621,6 +675,24 @@ function runFullASTValidation(
+          if (functions.length > requiredFunctions) {
+            errors.push(
+              getMessageFromId({
+                messageId: 'tooManyFirstArguments',
```
The message still says `a single ...`, but I think it's fine for now - as you mentioned the parsing logic can't handle this either at the moment.

wylieconlon reviewed (changes requested):
On a meta level, I think this code is hard to deal with because of `groupArgsByType`. For example `sum(bytes, 5)` is probably worth throwing an error for, same with `moving_average(sum(bytes), 5)`, but scanning through the code I'm not there's a simple validator for both. What I was _hoping_ for was a simple test that would catch all of the "extra arguments" cases, like what we do for Math: `if (node.args.length > positionalArguments.length) {`. So if I can suggest a refactoring here, it would be that I'd *iterate over each argument in position order* instead of using the grouped arguments. This would simplify the code and also catch edge cases like the one I've mentioned above.

dej611 commented:
> On a meta level, I think this code is hard to deal with because of `groupArgsByType`. For example `sum(bytes, 5)` is probably worth throwing an error for, same with `moving_average(sum(bytes), 5)`, but scanning through the code I'm not there's a simple validator for both. What I was _hoping_ for was a simple test that would catch all of the "extra arguments" cases, like what we do for Math: `if (node.args.length > positionalArguments.length) {`. So if I can suggest a refactoring here, it would be that I'd _iterate over each argument in position order_ instead of using the grouped arguments. This would simplify the code and also catch edge cases like the one I've mentioned above.

I think this makes sense over a general refactoring of the validation code, rather than this specific PR to fix some edge cases.

I've refactored the code now and added many more test cases with both positional and type-based issues and it all passes with the new implementation.

dej611 commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136791/)
* Commit: aa4b7f67e7809040b5490f22e20e157465f0697d
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104092/aa4b7f67e7809040b5490f22e20e157465f0697d)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/f9f5cf2249ecc5cc01654cc2a88a1a93381e6273) | [after](https://github.com/elastic/kibana/commit/aa4b7f67e7809040b5490f22e20e157465f0697d) | diff |
| --- | --- | --- | --- |
| `lens` | 1.5MB | 1.5MB | +4.1KB |

### History
* :green_heart: [Build #136321](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/) succeeded 974808337015f94ec1cc9c8e2422b8b77c5ac7bd
* :broken_heart: [Build #136294](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/) failed 950ff349365524c1982c14fe062a492af499d729
* :green_heart: [Build #136259](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/) succeeded 13eb57cfdeb019aaa92ef8162abe30cd64b960ba
* :green_heart: [Build #136221](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/) succeeded b50ecd55e0124313ad62b1c18db48cb62ed01852
* :green_heart: [Build #135961](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/) succeeded 0dcc0b83ead8cc942efa11b41bfe4b1c549794bd

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/135961/","number":"135961","commit":"0dcc0b83ead8cc942efa11b41bfe4b1c549794bd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/","number":"136221","commit":"b50ecd55e0124313ad62b1c18db48cb62ed01852"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/","number":"136259","commit":"13eb57cfdeb019aaa92ef8162abe30cd64b960ba"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/","number":"136294","commit":"950ff349365524c1982c14fe062a492af499d729"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/","number":"136321","commit":"974808337015f94ec1cc9c8e2422b8b77c5ac7bd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136791/","number":"136791","commit":"aa4b7f67e7809040b5490f22e20e157465f0697d"}]}
PIPELINE-->

dej611 commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137163/)
* Commit: 07fb0a09a14782afd7daed81dc335e08f2f83042
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104092/07fb0a09a14782afd7daed81dc335e08f2f83042)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/646f1d4076fd1ab2589c429f09966d4659be7b71) | [after](https://github.com/elastic/kibana/commit/07fb0a09a14782afd7daed81dc335e08f2f83042) | diff |
| --- | --- | --- | --- |
| `lens` | 1.5MB | 1.5MB | +4.1KB |

### History
* :green_heart: [Build #136791](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136791/) succeeded aa4b7f67e7809040b5490f22e20e157465f0697d
* :green_heart: [Build #136321](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/) succeeded 974808337015f94ec1cc9c8e2422b8b77c5ac7bd
* :broken_heart: [Build #136294](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/) failed 950ff349365524c1982c14fe062a492af499d729
* :green_heart: [Build #136259](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/) succeeded 13eb57cfdeb019aaa92ef8162abe30cd64b960ba
* :green_heart: [Build #136221](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/) succeeded b50ecd55e0124313ad62b1c18db48cb62ed01852

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136221/","number":"136221","commit":"b50ecd55e0124313ad62b1c18db48cb62ed01852"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/","number":"136259","commit":"13eb57cfdeb019aaa92ef8162abe30cd64b960ba"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/","number":"136294","commit":"950ff349365524c1982c14fe062a492af499d729"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/","number":"136321","commit":"974808337015f94ec1cc9c8e2422b8b77c5ac7bd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136791/","number":"136791","commit":"aa4b7f67e7809040b5490f22e20e157465f0697d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137163/","number":"137163","commit":"07fb0a09a14782afd7daed81dc335e08f2f83042"}]}
PIPELINE-->

dej611 commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137533/)
* Commit: 7f571bfeb8059c5dcb9afe746e7fbc389b5ad17c
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104092/7f571bfeb8059c5dcb9afe746e7fbc389b5ad17c)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/3e118c4e23d716d71bbbccd63beeab7da0c7a577) | [after](https://github.com/elastic/kibana/commit/7f571bfeb8059c5dcb9afe746e7fbc389b5ad17c) | diff |
| --- | --- | --- | --- |
| `lens` | 1.5MB | 1.5MB | +4.1KB |

### History
* :green_heart: [Build #137163](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137163/) succeeded 07fb0a09a14782afd7daed81dc335e08f2f83042
* :green_heart: [Build #136791](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136791/) succeeded aa4b7f67e7809040b5490f22e20e157465f0697d
* :green_heart: [Build #136321](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/) succeeded 974808337015f94ec1cc9c8e2422b8b77c5ac7bd
* :broken_heart: [Build #136294](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/) failed 950ff349365524c1982c14fe062a492af499d729
* :green_heart: [Build #136259](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/) succeeded 13eb57cfdeb019aaa92ef8162abe30cd64b960ba

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136259/","number":"136259","commit":"13eb57cfdeb019aaa92ef8162abe30cd64b960ba"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/","number":"136294","commit":"950ff349365524c1982c14fe062a492af499d729"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/","number":"136321","commit":"974808337015f94ec1cc9c8e2422b8b77c5ac7bd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136791/","number":"136791","commit":"aa4b7f67e7809040b5490f22e20e157465f0697d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137163/","number":"137163","commit":"07fb0a09a14782afd7daed81dc335e08f2f83042"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137533/","number":"137533","commit":"7f571bfeb8059c5dcb9afe746e7fbc389b5ad17c"}]}
PIPELINE-->

dej611 commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137762/)
* Commit: d863ca40162e0ab9844c62808f4811af01a9e602
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-104092/d863ca40162e0ab9844c62808f4811af01a9e602)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b2b57a240423e9503d1307075d34c73eb8b2a3c9) | [after](https://github.com/elastic/kibana/commit/d863ca40162e0ab9844c62808f4811af01a9e602) | diff |
| --- | --- | --- | --- |
| `lens` | 1.5MB | 1.5MB | +4.1KB |

### History
* :green_heart: [Build #137533](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137533/) succeeded 7f571bfeb8059c5dcb9afe746e7fbc389b5ad17c
* :green_heart: [Build #137163](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137163/) succeeded 07fb0a09a14782afd7daed81dc335e08f2f83042
* :green_heart: [Build #136791](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136791/) succeeded aa4b7f67e7809040b5490f22e20e157465f0697d
* :green_heart: [Build #136321](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/) succeeded 974808337015f94ec1cc9c8e2422b8b77c5ac7bd
* :broken_heart: [Build #136294](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/) failed 950ff349365524c1982c14fe062a492af499d729

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136294/","number":"136294","commit":"950ff349365524c1982c14fe062a492af499d729"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136321/","number":"136321","commit":"974808337015f94ec1cc9c8e2422b8b77c5ac7bd"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/136791/","number":"136791","commit":"aa4b7f67e7809040b5490f22e20e157465f0697d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137163/","number":"137163","commit":"07fb0a09a14782afd7daed81dc335e08f2f83042"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137533/","number":"137533","commit":"7f571bfeb8059c5dcb9afe746e7fbc389b5ad17c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/137762/","number":"137762","commit":"d863ca40162e0ab9844c62808f4811af01a9e602"}]}
PIPELINE-->

wylieconlon reviewed (approved):
Tested and LGTM. It's still missing the case where you do `sum(bytes, 5)` but this is worth getting into 7.14 and refactoring later as it significantly improves the semantics.

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.14](https://github.com/elastic/kibana/pull/105395)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/105395">](https://github.com/elastic/kibana/pull/105395) |
| ✅ |  [7.x](https://github.com/elastic/kibana/pull/105396)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/105396">](https://github.com/elastic/kibana/pull/105396) |

The backport PRs will be merged automatically after passing CI.
